Add support for indexing submodules without repocache

H3mul opened the pull request:
This PR aims to make git repo submodule indexing much more approachable, removing the requirement for a repo cache

Related discussion : https://github.com/sourcegraph/zoekt/issues/984

# Behavior changes:

- If `zoekt-git-index` is run with `-submodules=true` and `-repo-cache` is not empty, the behavior remains unchanged
- If `zoekt-git-index` is run with `-submodules=true` and `-repo-cache` is empty, recurse on submodules using `go-git` instead

This required 2 related bug fixes:

- https://github.com/sourcegraph/zoekt/commit/085b391077b2e90de6515f23c919abc390de50a9 fixes an issue where we incorrectly replicated `git.PlainOpen()` functionality when setting up filesystem storage. A side effect of this is that `go-git` refuses to return submodules (not sure if there are others).

  Side note: sent `go-git` a PR to avoid this code replication entirely: 
  https://github.com/go-git/go-git/pull/1686

- https://github.com/sourcegraph/zoekt/commit/f5f787720bee7947ae671accff567cb268d8afdb adds subrepository support to RepoURL and LineFragment treatment - without this, results involving subrepositories can't be used to construct correct remote URLs

# Functionality Example:

indexing [Chataigne](https://github.com/benkuper/Chataigne)

```
$ zoekt-git-index -submodules=true ~/dev/Chataigne

2025/10/08 12:21:31 adding subrepository files from: Modules/juce_timeline
2025/10/08 12:21:31 adding subrepository files from: asio
2025/10/08 12:21:31 adding subrepository files from: Modules/juce_simpleweb
2025/10/08 12:21:31 adding subrepository files from: Modules/juce_sharedtexture
2025/10/08 12:21:31 failed to open submodule repository: Modules/juce_serial, submodule not initialized
2025/10/08 12:21:31 adding subrepository files from: Modules/juce_dmx
2025/10/08 12:21:31 adding subrepository files from: Modules/juce_organicui
2025/10/08 12:21:31 attempting to index 3745 total files
2025/10/08 12:04:20 finished shard .zoekt/github.com%2Fbenkuper%2FChataigne_v16.00000.zoekt: 91189557 index bytes (overhead 2.7), 3745 files processed 
```

<details>
<summary>search query</summary>


```
$ curl -XPOST -d '{"Q":"juce"}' localhost:6070/api/search

...

     {
        "FileName": "Modules/juce_dmx/README.md",
        "Repository": "github.com/benkuper/Chataigne",
        "SubRepositoryName": "github.com/benkuper/juce_dmx",
        "SubRepositoryPath": "Modules/juce_dmx",
        "Version": "dba0680390873dd1a4b820aa77afff78d61894d9",
        "Language": "Markdown",
        "Branches": [
          "HEAD"
        ],
        "LineMatches": [

...

    "RepoURLs": {
      "github.com/benkuper/Chataigne": "{{URLJoinPath \"https://github.com/benkuper/Chataigne\" \"blob\" .Version .Path}}",
      "github.com/benkuper/asio": "{{URLJoinPath \"https://github.com/benkuper/asio\" \"blob\" .Version .Path}}",
      "github.com/benkuper/juce_dmx": "{{URLJoinPath \"https://github.com/benkuper/juce_dmx\" \"blob\" .Version .Path}}",
      "github.com/benkuper/juce_organicui": "{{URLJoinPath \"https://github.com/benkuper/juce_organicui\" \"blob\" .Version .Path}}",
      "github.com/benkuper/juce_sharedtexture": "{{URLJoinPath \"https://github.com/benkuper/juce_sharedtexture\" \"blob\" .Version .Path}}",
      "github.com/benkuper/juce_simpleweb": "{{URLJoinPath \"https://github.com/benkuper/juce_simpleweb\" \"blob\" .Version .Path}}",
      "github.com/benkuper/juce_timeline": "{{URLJoinPath \"https://github.com/benkuper/juce_timeline\" \"blob\" .Version .Path}}"
    },
    "LineFragments": {
      "github.com/benkuper/Chataigne": "#L{{.LineNumber}}",
      "github.com/benkuper/asio": "#L{{.LineNumber}}",
      "github.com/benkuper/juce_dmx": "#L{{.LineNumber}}",
      "github.com/benkuper/juce_organicui": "#L{{.LineNumber}}",
      "github.com/benkuper/juce_sharedtexture": "#L{{.LineNumber}}",
      "github.com/benkuper/juce_simpleweb": "#L{{.LineNumber}}",
      "github.com/benkuper/juce_timeline": "#L{{.LineNumber}}"
    }
...

```
</details>

<details>
<summary>repo list query</summary>

```
$ curl -XPOST -d '{}' localhost:6070/api/list

...
     {
        "Repository": {
          "TenantID": 0,
          "ID": 0,
          "Name": "github.com/benkuper/Chataigne",
          "URL": "https://github.com/benkuper/Chataigne",
          "Metadata": null,
          "Source": "/home/hemul/dev/Chataigne",
          "Branches": [
            {
              "Name": "HEAD",
              "Version": "e4f150fe5d7fde4c8a40937d3e91cfa1ba13accc"
            }
          ],
          "SubRepoMap": {
            "Modules/juce_dmx": {
              "TenantID": 0,
              "ID": 0,
              "Name": "github.com/benkuper/juce_dmx",
              "URL": "https://github.com/benkuper/juce_dmx",
              "Metadata": null,
              "Source": "",
              "Branches": [
                {
                  "Name": "HEAD",
                  "Version": "dba0680390873dd1a4b820aa77afff78d61894d9"
                }
              ],
              "SubRepoMap": null,
              "CommitURLTemplate": "{{URLJoinPath \"https://github.com/benkuper/juce_dmx\" \"commit\" .Version}}",
              "FileURLTemplate": "{{URLJoinPath \"https://github.com/benkuper/juce_dmx\" \"blob\" .Version .Path}}",
              "LineFragmentTemplate": "#L{{.LineNumber}}",
              "RawConfig": null,
              "Rank": 0,
              "IndexOptions": "",
              "HasSymbols": false,
              "Tombstone": false,
              "LatestCommitDate": "0001-01-01T00:00:00Z"
            },
            ...
```
</details>

# Followups

Some other features I'd like to follow up on in future PRs:

- Subrepo results are quite cumbersome: the submodule file path gets appended onto parent repo submodule path, this needs to be split to get a correct remote URL. It might be a good idea to add an option to index submodules as separate zoekt repos instead of subrepos, eg a `-submodule-treatment subrepository | repository (default: subrepository)` cli flag.
- supporting incremental/delta builds

H3mul committed:
Fix git.Open() fs argument for .git dir instead of worktree

The filesystem argument submitted to git.Open() is of the .git
directory, it should be for the worktree, as in go-git this logic is copied from:

https://github.com/go-git/go-git/blob/145daf2492dd86b1d7e7787555ebd9837b93bff8/repository.go#L373-L375

One side effect of this bug was that go-git would refuse to fetch repo
submodules

H3mul committed:
Fix missing RepoURLs and LineFragments for result subrepositories

RepoURLs for subrepositories are currently filtered out when results
are streamed per-repo from shards.

H3mul committed:
Add submodule indexing without repo cache

When not using repo cache, index git submodules recursively into

subrepos using the go-git API

keegancsmith reviewed (approved):
Looks good so approving! But I wanna just do a bit more review early next week before landing.

H3mul commented:
I took a look at the failing submodule test, it uncovered a bit of a further discussion on desired submodule treatment

Test repocache:

https://github.com/sourcegraph/zoekt/blob/e1cef6f2b4dc40a2075839273a729c26f9a61c27/gitindex/tree_test.go#L40

```
├── adir      #adir worktree
├── bdir      #bdir worktree
└── gerrit.googlesource.com   # repocache root
    ├── adir.git   # bare repo with submodule
    └── bdir.git   # bare repo 
```

```gitconfig
# gerrit.googlesource.com/adir.git/config
...
[remote "origin"]
    url = http://gerrit.googlesource.com/adir
```
```gitconfig
# gerrit.googlesource.com/bdir.git/config
...
[remote "origin"]
    url = /tmp/<temphash>/bdir
```
```gitconfig
# adir/.gitmodules
[submodule "bname"]
    path = bname
    url = ../bdir
```

When indexing `adir`, repo-cache treatment resolves the `bname` subrepository name AND remote URL to `gerrit.googlesource.com/bdir`, despite that remote never being listed anywhere, simply because `../bdir` is applied "relatively" to the `gerrit.googlesource.com/adir` remote.

https://github.com/sourcegraph/zoekt/blob/e1cef6f2b4dc40a2075839273a729c26f9a61c27/gitindex/tree.go#L53-L57

I didn't consider that submodules actually have 2 sources of urls: the url in `.gitmodule` and the url(s) in the submodule repo remote. The former can be accessed via `submodule.Config().URL`.

Imo, this should be ignored, and the submodule zoekt name/url should be treated the same as with parent repo:

- if it has a remote - use it for name and url
- if no remote - use dirname (or perhaps submodule path) for repo name and no url

However, this makes the behavior inconsistent with the repo-cache case. Thoughts?

keegancsmith commented:
Is this relative URL stuff a hack in zoekt, or is it matching the behaviour of git? If the former, I imagine this is very specific to indexing gerrit which was the original use case for zoekt. In practice I don't know if zoekt is actually used for gerrit these days. The search at https://cs.opensource.google/ is not using zoekt.

To be honest I agree with your opinion. I imagine the repo-cache should be setup so that this all just works? But yeah, I will rely a lot on your opinion here since I've never really touched this code in Zoekt + I don't like git submodules :P

H3mul committed:
Adjust subrepo naming conventions

- Revert repo-cache behavior to unchanged
- Add support for naming repos and modules with no remote (use the dirname for the
  former and subrepopath for the latter)
- Add test for submodule no-repo-cache behavior

H3mul committed:
Fix overallocation in repoURL and lineFragment filtering

H3mul commented:
>Is this relative URL stuff a hack in zoekt, or is it matching the behaviour of git?

Looks like a gerrit-specific treatment to me, yeah. Git only supports EITHER relative paths OR urls, doesn't mix them together.

Sounds like the repo-cache treatment has a specific use-case that should remain consistent, while the behavior this PR is introducing can make other choices.

I implemented the submodule naming scheme I proposed, and committed the overallocation fix we discussed

keegancsmith commented:
Sounds great. Mind working on getting `go test ./...` to pass then will review and merge :)

H3mul committed:
Revert basename treatment from 08f67597

Missed the fact that it is already done in zoekt-git-index
command, the test was improperly configured

https://github.com/sourcegraph/zoekt/blob/4e4a529c3b63c7d4c7897ba736f1cd52cc163134/cmd/zoekt-git-index/main.go#L93-L100

H3mul commented:
Should pass now with this last change

keegancsmith commented:
Thanks so much for the high quality PR!
